File: Medium/133-clone-graph.py
# Runtime: 77 ms, faster than 10.28% of Python3 online submissions for Clone Graph.
# Memory Usage: 14.4 MB, less than 83.99% of Python3 online submissions for Clone Graph.

"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

import collections

class Solution:
    def __init__(self):
        # initialize an empty dict of, which will store nodes & their copies.
        self.created = {}
    
    def cloneGraph(self, node: 'Node') -> 'Node':
        
        # Go through all nbors of node
        # create the appropriate new nodes,
        # then recursively call cloneGraph on those nodes
        # return node
        
        if node == None:
            return node
        
        if node in self.created.keys():
            return self.created[node]
        else:
            new = Node(val=node.val, neighbors=[])
            self.created[node] = new
            
        for nbor in node.neighbors:
            new.neighbors.append(self.cloneGraph(nbor))
                 
        return new
        
    # adj list = [[2, 4], [1, 3], [2, 4], [1, 3]]

File: Medium/133-clonegraph.cpp
#include <vector>
#include <unordered_map>

using namespace std;

/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* cgHelper(Node* node, unordered_map<int, Node*>& clones) {
        if (node == nullptr) {
            return nullptr;
        } else if (clones.find(node->val) != clones.end()) {
            // Node has already been seen
            return clones[node->val];
        }
        
        Node* clone = new Node(node->val);

        clones[node->val] = clone;
        
        for (Node* nbor: node->neighbors) {
            clone->neighbors.push_back(cgHelper(nbor, clones));
        }
        
        return clone;
    }
    
    Node* cloneGraph(Node* node) {        
        unordered_map<int, Node*> clones;
        
        return cgHelper(node, clones);
    }
};